packages/xpetra/src/Map/Xpetra_TpetraMap_def.hpp

```diff
@@ -160,260 +156,157 @@ namespace Xpetra {
-    RCP<const Map<LocalOrdinal, GlobalOrdinal, Node> > removeEmptyProcesses () const {
-      return toXpetra(map_->removeEmptyProcesses());
-    }
-    RCP<const Map<LocalOrdinal, GlobalOrdinal, Node> > replaceCommWithSubset (const Teuchos::RCP<const Teuchos::Comm<int> >& newComm) const {
-      return toXpetra(map_->replaceCommWithSubset(newComm));
-    }
+template<class LocalOrdinal, class GlobalOrdinal, class Node>
+RCP<const Map<LocalOrdinal, GlobalOrdinal, Node> > TpetraMap<LocalOrdinal,GlobalOrdinal,Node>::replaceCommWithSubset (const Teuchos::RCP<const Teuchos::Comm<int> >& newComm) const
+{
+    return toXpetra(map_->replaceCommWithSubset(newComm));
+}
 
 #ifdef XPETRA_ENABLE_DEPRECATED_CODE
-    template<class Node2>
-    RCP<Map<LocalOrdinal, GlobalOrdinal, Node2> > XPETRA_DEPRECATED clone(const RCP<Node2> &node2) const {
-      return toXpetraNonConst(map_->clone(node2));
-    }
+template<class LocalOrdinal, class GlobalOrdinal, class Node>
+template<class Node2>
+RCP<Map<LocalOrdinal, GlobalOrdinal, Node2> > XPETRA_DEPRECATED TpetraMap<LocalOrdinal,GlobalOrdinal,Node>::clone(const RCP<Node2> &node2) const
+{
+    return toXpetraNonConst(map_->clone(node2));
+}
 #endif
 
-    //@}
+template<class LocalOrdinal, class GlobalOrdinal, class Node>
+TpetraMap<LocalOrdinal,GlobalOrdinal,Node>::TpetraMap(const Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node > > &map)
+: map_(map) { }
 
-    //! @name Xpetra specific
-    //@{
+template<class LocalOrdinal, class GlobalOrdinal, class Node>
+UnderlyingLib lib() { return UseTpetra; }
+//! Get the library used by this TpetraMap<LocalOrdinal,GlobalOrdinal,Node>::object (Tpetra or Epetra?)    UnderlyingLib lib() const { return UseTpetra; }
+//! Get the library used by this object (Tpetra or Epetra?)    UnderlyingLib lib() const { return UseTpetra; }    UnderlyingLib lib() const { return UseTpetra; }
 
-    //! TpetraMap constructor to wrap a Tpetra::Map object
-    TpetraMap(const Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node > > &map)
-      : map_(map) { }
+template<class LocalOrdinal, class GlobalOrdinal, class Node>
+RCP< const Tpetra::Map< LocalOrdinal, GlobalOrdinal, Node > > TpetraMap<LocalOrdinal,GlobalOrdinal,Node>::getTpetra_Map() const
+{ return map_; }
 
-    //! Get the library used by this object (Tpetra or Epetra?)
-    UnderlyingLib lib() const { return UseTpetra; }
-
-    //! Get the underlying Tpetra map
-    RCP< const Tpetra::Map< LocalOrdinal, GlobalOrdinal, Node > > getTpetra_Map() const { return map_; }
 
 #ifdef HAVE_XPETRA_KOKKOS_REFACTOR
 #ifdef HAVE_XPETRA_TPETRA
-    using local_map_type = typename Map<LocalOrdinal, GlobalOrdinal, Node>::local_map_type;
-    /// \brief Get the local Map for Kokkos kernels.
-    local_map_type getLocalMap () const {
-      return map_->getLocalMap();
-    }
+
+template<class LocalOrdinal, class GlobalOrdinal, class Node>
+local_map_type TpetraMap<LocalOrdinal, GlobalOrdinal, Node>::getLocalMap () const
```

Comment:
@csiefer2 I'm seeing compilation errors here in the **def** file:

```
[ 58%] Building CXX object packages/xpetra/src/CMakeFiles/xpetra.dir/Utils/ExplicitInstantiation/Xpetra_TpetraMap.cpp.o
In file included from /ascldap/users/wcmclen/dev/trilinos-dev/build/muelu/packages/xpetra/src/Utils/ExplicitInstantiation/Xpetra_TpetraMap.cpp:55:0:
/ascldap/users/wcmclen/dev/trilinos-dev/source/Trilinos/packages/xpetra/src/Map/Xpetra_TpetraMap_def.hpp:303:1: error: ‘local_map_type’ does not name a type; did you mean ‘localtime’?
 local_map_type TpetraMap<LocalOrdinal, GlobalOrdinal, Node>::getLocalMap () const
 ^~~~~~~~~~~~~~
```

There is a `using local_map_type = typename Map<LocalOrdinal, GlobalOrdinal, Node>::local_map_type;` in the decl before the function declaration, so I'm not sure right now what we're missing.

Reply:
You can't use decl class member typedefs as return values in the def.

Just expand it.

Reply:
@csiefer2 :racehorse: :racehorse: :racehorse: